### PR TITLE
Support env vars in requests to spawn kernels

### LINKED ISSUE #128: Pass environment metadata to kernels on POST /api/kernels
Make it possible for clients to pass additional metadata through when creating kernels and have that metadata show up in the kernel environment. 

Use case: passing secrets to kernels from clients without having those secrets hardcoded at KG start or having to pass them via code execution on the websocket.

For example:

```
POST /api/kernels

{
  "name": "python3",
  "env": {
    "APIKEY" : "123456"
  }
}
```

The notebook API spec does not support the "env" key. I'm proposing extending it.

https://github.com/jupyter/notebook/blob/master/notebook/services/api/api.yaml#L454

If this isn't acceptable, it can also be implemented via headers which is less than ideal but keeps 100% compatibility with the notebook API spec.

### PR DESCRIPTION
**Pull request by parente**
* Hack around lack of access to the start_kernel method in the base class with an override
* Whitelist KERNEL_ prefixed vars, blacklist everything else
* Add test case

Fixes #128

### CODE REVIEW

**Comment by parente** [comment]
@minrk @rgbkrk Would appreciate your thoughts on this. The `partial` in there is a complete hack, but it felt better than reimplementing the entire base class method just to pass additional args to `km.start_kernel`. 

Could be we want to improve this in the notebook server to make the `start_kernel` easier to customize. Could also be that this is a place where the dependency inversion of KG -> notebook is shining through.

File: kernel_gateway/services/kernels/handlers.py [hunk 1]
```diff
@@ -38,7 +40,23 @@ def post(self):
...
+        model = self.get_json_body()
+        if model is not None and 'env' in model:
+            if not isinstance(model['env'], dict):
+                raise tornado.web.HTTPError(400)
+            # Whitelist to KERNEL_* args only
+            env = {key: value for key, value in model['env'].items() if key.startswith('KERNEL_')}
+            # No way to override the call to start_kernel on the kernel manager
+            # so do a temporary partial (ugh)
+            orig_start = self.kernel_manager.start_kernel
+            self.kernel_manager.start_kernel = partial(self.kernel_manager.start_kernel, env=env)
```
Comment (rgbkrk) [inline kernel_gateway/services/kernels/handlers.py]: I'm partial to partials, though that's my pseudo-FP hat being on.

**Comment by parente** [comment]
The other elephant in the room here is about the API spec. This technically breaks compatibility with the notebook REST API spec, even though the `env` object is optional. How should we manage the spec and these proposals going forward? I don't think anyone wants to see the notebook and KG APIs grow far apart.

**Comment by rgbkrk** [comment]
I'd love to explore the API with those interested during the in-person dev meeting, even if we go ahead with this.

**Comment by minrk** [comment]
I think most proposals like this one should be considered as a testing ground for changes to the upstream API. For instance, I see no reason why something like this would be unique to KG, so pushing it upstream once it's ironed out here and we decide people like it would make sense. That's just my opinion, of course.

:+1: to chatting about this at the dev meeting.

**Comment by parente** [comment]
Relevant: @Lull3rSkat3r is working on exposing a swagger spec under the same route as in the notebook server here for #135. One line of thinking goes, as long as the API is documented, then there's really no harm in the KG and NB server APIs diverging slightly / temporarily.

**Comment by rgbkrk** [comment]
Does the code-quality/landscape check ever finish?

**Comment by parente** [comment]
It has on all the other PRs. @Carreau set it up so I'm not sure I can kick it or whatever.

I'm not in a rush to merge this. Was just making a note for a chat next week.

**Comment by parente** [comment]
Or we can ship it. Sure! :package:

**Comment by rgbkrk** [comment]
:grin: Right after I clicked the green button I saw "not in a rush".

**Pull request merged**